src/soter/openssl/soter_ec_key.c

```diff
@@ -14,6 +14,8 @@
  * limitations under the License.
  */
 
+#ifndef THEMIS_EXPERIMENTAL_OPENSSL_3_SUPPORT
```

Comment:
Instead of ifdeffing the entire file, shouldn't this be handled by Makefile instead?

Compile and link `openssl/soter_ec_key.c` when building for OpenSSL 1.0.2–1.1.1, or `openssl/soter_ec_key_3.c` when building for OpenSSL 3.

Reply:
imho, much better to use one approach instead of several. If we use somewhere definitions, so let use them everywhere instead of propagation that logic to another mechanism. Or we switch in the Makefile, or via definition passed to compiler. IMHO, second approach is better because it easier to re-use in the other build systems- and doesn't require other users to read and explore Makefile which files related to the preferred option.

Reply:
Regarding second approach, conditional selection of files to compile
* Will need to replace `CRYPTO_ENGINE_SOURCES += $(wildcard $(CRYPTO_ENGINE)/*.c)` in make rules with something more specific. Maybe list all files explicitly, and then, based on build option, append OpenSSL 1 or OpenSSL 3 specific files. Maybe use wildcard and apply some filter over it (if condition, remove these files, otherwise remove those ones)
* There are still few places where only a couple of rows in file need to be changed, so I don't think copying such files is a good idea, so the approach with `#ifdef` will still exist in such places
* Will need to update `CMakeLists.txt` as well OR just drop it. Nowadays IDEs like CLion have better support for Makefile projects, not sure about debugging capabilities though, gotta check how to launch debug targets

Reply:
If you're not using the Makefile, you still need to decide where OpenSSL is located, which one you want to use if you have multiple, and know how to compile Themis with it.

If anything, I'd auto-detect OpenSSL version using OPENSSL_VERSION_NUMBER macro (or any of its friends). Then we don't need our own. Whatever you end up #include'ing is the one you're compiling with anyway.